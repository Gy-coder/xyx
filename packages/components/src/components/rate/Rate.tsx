import {
  forwardRef,
  ForwardRefRenderFunction,
  useCallback,
  useState,
} from "react";
import classnames from "classnames";
import { RateProps } from "./interface";
import useMergeState from "../../hooks/useMergeState";
import "./index.scss";
import InternalIcon from "../_interal/internal_icon/InternalIcon";
import RateContext from "./context";
import RateItem from "./RateItem";

const Rate: ForwardRefRenderFunction<any, RateProps> = (props, ref) => {
  const {
    defaultValue,
    value,
    onChange,
    onHoverChange,
    character = <InternalIcon name="icon-star-fill" />,
    disabled = false,
    clearable = false,
    count = 5,
    half = false,
    size,
    className,
    style,
    hint,
  } = props;
  const [stateValue, setStateValue, isControlled] = useMergeState<number>(0, {
    defaultValue,
    value,
  });

  const [hover, setHover] = useState(0);

  const handleChange = useCallback(
    (v: number) => {
      if (disabled) return;
      const newValue = clearable && v === stateValue ? 0 : v;
      if (!isControlled) setStateValue(newValue);
      onChange?.(newValue);
    },
    [stateValue, clearable]
  );

  const handleMouseEnter = (v: number) => {
    if (disabled) return;
    setHover(v);
    onHoverChange?.(v);
  };
  const handleMouseLeave = () => {
    if (disabled) return;
    setHover(0);
  };
  if (hint && hint.length !== count)
    throw new Error("hint.length must be equal with count");
  return (
    <RateContext.Provider
      value={{
        stateValue,
        hover,
        disabled,
        size,
        onChange: handleChange,
        onMouseEnter: handleMouseEnter,
        onMouseLeave: handleMouseLeave,
      }}
    >
      <div
        className={classnames("g-rate", className, {
          "g-rate-disabled": disabled,
          [`g-rate-${size}`]: size,
        })}
        style={style}
        ref={ref}
      >
        {new Array(count)
          .fill(false)
          .map((_, i) => i + 1)
          .map((item) => {
            return (
              <span className="g-rate-item-wrapper" key={item}>
                {half && <RateItem value={item} character={character} half />}
                <RateItem value={item} character={character} />
              </span>
            );
          })}
        {hint && <span className="g-rate-hint">{hint[stateValue - 1]}</span>}
      </div>
    </RateContext.Provider>
  );
};

export default forwardRef(Rate);
