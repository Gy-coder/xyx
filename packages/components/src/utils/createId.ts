let id = 0;

export default function createId(){
    return id++
}