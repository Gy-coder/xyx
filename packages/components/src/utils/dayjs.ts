type DayUnit = "year"
    | "month"
    | "day"
    | "hour"
    | "minute"
    | "second"
    | "millisecond";

export default class Dayjs {
    public date: Date

    constructor(date?: number | string | Date) {
        this.date = date ? new Date(date) : new Date()
    }

    get year() {
        return this.date.getFullYear()
    }

    get month() {
        return this.date.getMonth() + 1
    }

    get week() {
        return this.date.getDay()
    }

    get day() {
        return this.date.getDate()
    }

    get hour() {
        return this.date.getHours()
    }

    get minute() {
        return this.date.getMinutes()
    }

    get second() {
        return this.date.getSeconds()
    }

    get millisecond() {
        return this.date.getMilliseconds()
    }

    get raw() {
        return this.date
    }

    public firstDayOfMonth() {
        return new Dayjs(
            new Date(
                this.year,
                this.month - 1,
                1,
                0,
                0,
                0)
        )
    }

    public add(amount: number, unit: DayUnit) {
        let date = new Date(this.date.getTime());
        switch (unit) {
            case 'year':
                const currentDate = date.getDate()
                date.setDate(1)
                date.setFullYear(date.getFullYear() + amount)
                const targetDate = new Date(
                    date.getFullYear(),
                    date.getMonth() + 1,
                    0,
                    0,
                    0,
                    0,
                ).getDate()
                date.setDate(Math.min(currentDate, targetDate))
                break;
            case 'month':
                const d = date.getDate()
                date.setDate(1)
                date.setMonth(date.getMonth() + amount);
                const d2 = new Date(date.getFullYear(), date.getMonth() + 1, 0, 0, 0, 0).getDate()
                date.setDate(Math.min(d, d2))
                break;
            case 'day':
                date.setDate(date.getDate() + amount);
                break;
            case 'hour':
                date.setHours(date.getHours() + amount);
                break;
            case 'minute':
                date.setMinutes(date.getMinutes() + amount);
                break;
            case 'second':
                date.setSeconds(date.getSeconds() + amount);
                break;
            case 'millisecond':
                date.setMilliseconds(date.getMilliseconds() + amount);
                break;

            default:
                throw new Error('Time.add: unknown unit');
        }
        return new Dayjs(date)
    }

    public set(amount: number,unit: DayUnit){
        const date = new Date(this.date.getTime())
        switch (unit){
            case "year":
                date.setFullYear(amount)
                break
            case "month":
                date.setMonth(amount - 1)
                break
            case "day":
                date.setDate(amount)
                break
            case "hour":
                date.setHours(amount)
                break
            case "second":
                date.setSeconds(amount)
                break
            case "millisecond":
                date.setMilliseconds(amount)
                break
            default:
                throw new Error("Time.add: unknown unit")
        }
        return new Dayjs(date)
    }

    isSameDay(day: Dayjs){
        return day.year === this.year && day.month === this.month && day.day === this.day
    }

    isSameMonth(day: Dayjs){
        return day.year === this.year && day.month === this.month
    }

    //@ts-ignore
    format(pattern?: string = 'YYYY-MM-DD') {
        // 目前支持的格式有 YYYY MM DD HH mm ss SSS
        return pattern.replace(/YYYY/g, this.year.toString())
            .replace(/MM/, this.month.toString().padStart(2, '0'))
            .replace(/DD/, this.day.toString().padStart(2, '0'))
            .replace(/HH/, this.hour.toString().padStart(2, '0'))
            .replace(/mm/, this.minute.toString().padStart(2, '0'))
            .replace(/ss/, this.second.toString().padStart(2, '0'))
            .replace(/SSS/, this.millisecond.toString().padStart(3, '0'))
    }
}
