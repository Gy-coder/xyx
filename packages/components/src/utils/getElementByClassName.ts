export function getElementByClassName(container = document, classname: string) {
  return container.getElementsByClassName(classname)[0];
}
