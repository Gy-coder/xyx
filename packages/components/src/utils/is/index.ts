export { default as isUndefined } from "./isUndefined";
export { default as isFunction } from "./isFunction";
