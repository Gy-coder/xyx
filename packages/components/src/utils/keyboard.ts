export const keyboard = {
  Enter: "Enter",
};
