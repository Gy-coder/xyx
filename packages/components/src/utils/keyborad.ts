export const keyborad = {
  Enter: "Enter",
};
