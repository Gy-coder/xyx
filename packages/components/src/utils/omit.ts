export default function omit<T = any>(obj: Record<string, T>, paths: string[]) {
  const res: Record<string, T> = { ...obj },
    set = new Set(paths);
  for (let key in obj) {
    if (set.has(key)) delete res[key];
  }
  return res;
}
