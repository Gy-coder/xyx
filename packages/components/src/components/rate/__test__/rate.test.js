import { fireEvent, render } from "@testing-library/react";
import "@testing-library/jest-dom";
import { getElementByClassName } from "../../../utils/getElementByClassName";
import Rate from "../Rate";
import { useState } from "react";

const rateClass = "g-rate";
const starClass = "g-rate-item";
const halfStarClass = "g-rate-item-half";
const starActive = "g-rate-item-active";
const disabledClass = "g-rate-disabled";
const disabledCItemClass = "g-rate-item-disabled";
const smallClass = "g-rate-small";
const smallItemClass = "g-rate-item-small";
const largeClass = "g-rate-large";
const largeItemClass = "g-rate-item-large";

describe("Rate component test", () => {
  it("can render corrently", () => {
    const { container, asFragment } = render(<Rate />);
    const el = getElementByClassName(container, rateClass);
    expect(el).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });
  it("add ClassName", () => {
    const { container } = render(<Rate className="123" />);
    const el = getElementByClassName(container, rateClass);
    expect(el).toHaveClass("123");
  });
  it("value", () => {
    const active = 3;
    const { container } = render(<Rate defaultValue={active} />);
    const stars = container.getElementsByClassName(starClass);
    for (let i = 0; i < active; i++) expect(stars[i]).toHaveClass(starActive);
    for (let i = active; i < stars.length; i++)
      expect(stars[i]).not.toHaveClass(starActive);
  });
  it("can change value", () => {
    const jestFn = jest.fn();
    const { container } = render(<Rate onChange={jestFn} />);
    const stars = container.getElementsByClassName(starClass);
    for (let i = 0; i < stars.length; i++) expect(stars[i]).toBeVisible();
    fireEvent.click(stars[3]);
    expect(jestFn).toBeCalledWith(4);
    fireEvent.click(stars[0]);
    expect(jestFn).toBeCalledWith(1);
  });
  it("count", () => {
    const { container } = render(<Rate count={10} />);
    const stars = container.getElementsByClassName(starClass);
    expect(stars.length).toBe(10);
  });
  it("can clear value", () => {
    const jestFn = jest.fn();
    const { container } = render(
      <Rate onChange={jestFn} defaultValue={3} clearable />
    );
    const stars = container.getElementsByClassName(starClass);
    const star = container.getElementsByClassName(starActive)[2];
    fireEvent.click(star);
    expect(jestFn).toBeCalledWith(0);
    for (let i = 0; i < star.length; i++)
      expect(stars[i]).not.toHaveClass(starActive);
    fireEvent.click(stars[2]);
    expect(jestFn).toBeCalledWith(3);
    for (let i = 0; i < 3; i++) expect(stars[i]).toHaveClass(starActive);
    fireEvent.click(stars[2]);
    expect(jestFn).toBeCalledWith(0);
    for (let i = 0; i < star.length; i++)
      expect(stars[i]).not.toHaveClass(starActive);
  });
  it("controlled component", () => {
    const jestFn = jest.fn();
    let v, setV, handleChange;
    const Demo = () => {
      [v, setV] = useState(0);
      handleChange = (newValue) => {
        setV(newValue);
        jestFn(newValue);
      };
      return <Rate value={v} onChange={handleChange} />;
    };
    const { container } = render(<Demo />);
    const stars = container.getElementsByClassName(starClass);
    fireEvent.click(stars[4]);
    expect(v).toBe(5);
    fireEvent.click(stars[1]);
    expect(v).toBe(2);
  });
  it("test disabled", () => {
    const jestFn = jest.fn();
    const { container } = render(
      <Rate onChange={jestFn} defaultValue={3} disabled />
    );
    const el = getElementByClassName(container, rateClass);
    expect(el).toHaveClass(disabledClass);
    const stars = container.getElementsByClassName(starClass);
    for (let i = 0; i < stars.length; i++) {
      fireEvent.click(stars[i]);
      expect(jestFn).not.toBeCalled();
    }
  });
  it("charactor", () => {
    const charactor = 1;
    const { container } = render(<Rate character={charactor} />);
    const stars = container.getElementsByClassName(starClass);
    for (let i = 0; i < stars.length; i++) expect(stars[i].innerHTML).toBe("1");
  });
  it("half", () => {
    const jestFn = jest.fn();
    const { container } = render(<Rate half onChange={jestFn} />);
    const halfStars = container.getElementsByClassName(halfStarClass);
    for (let i = 0; i < halfStars.length; i++)
      expect(halfStars[i]).toBeVisible();
    fireEvent.click(halfStars[0]);
    expect(jestFn).toBeCalledWith(0.5);
  });
  it("disabled", () => {
    const jestFn = jest.fn(),
      jestFn2 = jest.fn();
    const { container } = render(
      <Rate disabled onChange={jestFn} onHoverChange={jestFn2} />
    );
    const el = getElementByClassName(container, rateClass);
    expect(el).toBeVisible();
    expect(el).toHaveClass(disabledClass);
    const stars = container.getElementsByClassName(disabledCItemClass);
    expect(stars.length).toBe(5);
    for (let i = 0; i < stars.length; i++) {
      fireEvent.click(stars[i]);
      expect(jestFn).not.toBeCalled();
    }
    for (let i = 0; i < stars.length; i++) {
      fireEvent.mouseEnter(stars[i]);
      expect(jestFn2).not.toBeCalled();
    }
  });
  it("size small", () => {
    const { container } = render(<Rate size="small" />);
    const el = getElementByClassName(container, rateClass);
    expect(el).toBeVisible();
    expect(el).toHaveClass(smallClass);
    const stars = container.getElementsByClassName(smallItemClass);
    for (let i = 0; i < stars.length; i++) {
      expect(stars[i]).toHaveClass(smallItemClass);
    }
  });
  it("size large", () => {
    const { container } = render(<Rate size="large" />);
    const el = getElementByClassName(container, rateClass);
    expect(el).toBeVisible();
    expect(el).toHaveClass(largeClass);
    const stars = container.getElementsByClassName(largeItemClass);
    for (let i = 0; i < stars.length; i++) {
      expect(stars[i]).toHaveClass(largeItemClass);
    }
  });
});
