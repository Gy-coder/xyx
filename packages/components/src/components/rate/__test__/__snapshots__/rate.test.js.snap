// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rate component test can render corrently 1`] = `
<DocumentFragment>
  <div
    class="g-rate"
  >
    <span
      class="g-rate-item-wrapper"
    >
      <span
        class="g-rate-item"
        tabindex="0"
      >
        <svg
          class="icon"
        >
          <use
            xlink:href="#icon-star-fill"
          />
        </svg>
      </span>
    </span>
    <span
      class="g-rate-item-wrapper"
    >
      <span
        class="g-rate-item"
        tabindex="0"
      >
        <svg
          class="icon"
        >
          <use
            xlink:href="#icon-star-fill"
          />
        </svg>
      </span>
    </span>
    <span
      class="g-rate-item-wrapper"
    >
      <span
        class="g-rate-item"
        tabindex="0"
      >
        <svg
          class="icon"
        >
          <use
            xlink:href="#icon-star-fill"
          />
        </svg>
      </span>
    </span>
    <span
      class="g-rate-item-wrapper"
    >
      <span
        class="g-rate-item"
        tabindex="0"
      >
        <svg
          class="icon"
        >
          <use
            xlink:href="#icon-star-fill"
          />
        </svg>
      </span>
    </span>
    <span
      class="g-rate-item-wrapper"
    >
      <span
        class="g-rate-item"
        tabindex="0"
      >
        <svg
          class="icon"
        >
          <use
            xlink:href="#icon-star-fill"
          />
        </svg>
      </span>
    </span>
  </div>
</DocumentFragment>
`;
